lib/content-services/src/lib/infinite-scroll-datasource/infinite-scroll-datasource.ts

```diff
@@ -0,0 +1,82 @@
+/*!
+ * @license
+ * Copyright © 2005-2023 Hyland Software, Inc. and its affiliates. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { ContentPagingQuery } from '@alfresco/js-api';
+import { CollectionViewer, DataSource } from '@angular/cdk/collections';
+import { BehaviorSubject, forkJoin, Observable, Subject, Subscription } from 'rxjs';
+import { take, tap } from 'rxjs/operators';
+
+export abstract class InfiniteScrollDatasource<T> extends DataSource<T> {
+    protected readonly dataStream = new BehaviorSubject<T[]>([]);
+    private isLoading$ = new Subject<boolean>();
+    private subscription = new Subscription();
+    private batchesFetched = 0;
+    private _itemsCount = 0;
+    private _firstItem: T;
+
+    /* Determines size of each batch to be fetched */
+    batchSize = 100;
+
+    /* Observable with initial and on reset loading state */
+    isLoading = this.isLoading$.asObservable();
+
+    get itemsCount(): number {
+        return this._itemsCount;
+    }
+
+    get firstItem(): T {
+        return this._firstItem;
+    }
+
+    abstract getNextBatch(pagingOptions: ContentPagingQuery): Observable<T[]>;
+
+    connect(collectionViewer: CollectionViewer): Observable<T[]> {
+        this.reset();
```

Comment:
In previous commit you had here this.batchesFetched += 1. Now you use reset function (what is correct) but that function has this.batchesFetched = 1. Is that correct? Maybe we should parametrize function or set this outside function if we need different value?

Reply:
Yes I thought it through and since `connect` is called only once at the beginning it's okay to have `= 1` there as we will have only one batch fetched there at the time